stdnum/isl/kennitala.py

```diff
@@ -0,0 +1,116 @@
+# kennitala.py - functions for handling Icelandic identity codes
+# coding: utf-8
+#
+# Copyright (C) 2011 Jussi Judin
+# Copyright (C) 2012, 2013 Arthur de Jong
+#
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301 USA
+
+"""Kennitala (Icelandic personal and organisation identity code).
+
+Module for handling Icelandic personal and organisation identity codes
+(kennitala).
+
+>>> validate('530575-0209')  # organisation
+'530575-0209'
+>>> validate('120174-3399')  # individual
+'1201743399'
+>>> validate('530575-0299')
+Traceback (most recent call last):
+    ...
+InvalidChecksum: ...
+>>> validate('320174-3399')
+Traceback (most recent call last):
+    ...
+InvalidComponent: ...
+>>> compact('530575-0209')
+'5305750209'
+"""
+
+import re
+import datetime
+
+from stdnum.exceptions import *
+from stdnum.util import clean
+
+
+# Icelandic personal and organisation identity codes are composed of
+# date part, a dash, two random digits, a checksum, and a century
+# indicator where '9' for 1900-1999 and '0' for 2000 and beyond. For
+# organisations instead of birth date, the registration date is used,
+# and number 4 is added to the first digit.
+_kennitala_re = re.compile(r'^(?P<day>[01234567]\d)(?P<month>[01]\d)(?P<year>\d\d)'
+                           r'(?P<random>\d\d)(?P<control>\d)'
+                           r'(?P<century>[09])$')
+
+
+def compact(number):
+    """Convert the kennitala to the minimal representation. This
+    strips surrounding whitespace and separation dash, and converts it
+    to upper case."""
+    return clean(number, '-').upper().strip()
```

Comment:
@toivotuo looks quite strange when upper is applied to number :) though no harm there